x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Intersects.java

```diff
@@ -0,0 +1,234 @@
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.Line;
+import org.elasticsearch.geometry.Point;
+import org.elasticsearch.geometry.Polygon;
+import org.elasticsearch.geometry.Rectangle;
+import org.elasticsearch.xpack.esql.planner.Mappable;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.TypeResolutions;
+import org.elasticsearch.xpack.ql.expression.function.scalar.BinaryScalarFunction;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+import org.elasticsearch.xpack.ql.util.SpatialUtils;
+
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.ql.type.DataTypes.GEO_POINT;
+import static org.elasticsearch.xpack.ql.type.DataTypes.isString;
+
+public class Intersects extends BinaryScalarFunction implements Mappable {
+
+    public Intersects(Source source, Expression timestamp, Expression argument) {
+        super(source, timestamp, argument);
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.BOOLEAN;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        TypeResolution resolution = isGeoPointOrString(left(), sourceText(), FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        resolution = isGeoPointOrString(right(), sourceText(), SECOND);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    public static TypeResolution isGeoPointOrString(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
+        return isType(e, dt -> dt == GEO_POINT || isString(dt), operationName, paramOrd, "geo_point", "keyword", "text");
+    }
+
+    @Override
+    protected Intersects replaceChildren(Expression newLeft, Expression newRight) {
+        return new Intersects(source(), newLeft, newRight);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, Intersects::new, left(), right());
+    }
+
+    @Override
+    public boolean foldable() {
+        return left().foldable() && right().foldable();
+    }
+
+    @Override
+    public Object fold() {
+        return Mappable.super.fold();
+    }
+
+    @Override
+    public Supplier<EvalOperator.ExpressionEvaluator> toEvaluator(
+        Function<Expression, Supplier<EvalOperator.ExpressionEvaluator>> toEvaluator
+    ) {
+        if (isString(left().dataType())) {
```

Comment:
I suggest some kind of casting infrastructure to prevent the combinatorial explosion. `Cast#cast` has some of this for numbers. You'd want some for your spatial stuff.

Right now the casting has a cost - we cast block at a time  and your evaluators can do this value at a time - but we won't be doing that forever. I want runtime code generation, especially for casts, and that'd help there.